File: scribly/sendgrid.py
import logging

import aiohttp

from scribly.definitions import Email, EmailGateway

logger = logging.getLogger(__name__)


class SendGrid(EmailGateway):
    sender = "emails@scribly.app"

    def __init__(
        self, api_key: str, base_url: str, session: aiohttp.ClientSession
    ) -> None:
        self.session = session
        self.api_key = api_key
        self.base_url = base_url

    async def send_email(self, email: Email) -> None:
        logger.info("Sending email %s", email)
        url = f"{self.base_url}/v3/mail/send"
        body = {
            "personalizations": [
                {
                    "to": [{"email": email.to.email, "name": email.to.username}],
                    "subject": email.subject,
                }
            ],
            "from": {"email": self.sender},
            "content": [{"type": "text/html", "value": email.body}],
        }
        headers = {"Authorization": f"Bearer {self.api_key}"}
        async with self.session.post(url, json=body, headers=headers) as response:
            if response.status < 200 or response.status > 299:
                text = await response.text()
                raise RuntimeError(
                    f"Received non-200 response from sendgrid {response.status} - '{text}'."
                )

File: internal/sendgrid.go
package internal

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
)

type SendgridClient struct {
	client  *http.Client
	baseURL string
	apiKey  string
}

func (s *SendgridClient) SendEmail(ctx context.Context, email Email) error {
	sendgridEmail := map[string]interface{}{
		"personalizations": []map[string]interface{}{
			{
				"to": []map[string]interface{}{
					{
						"email": email.To.Email,
						"name":  email.To.Username,
					},
				},
				"subject": email.Subject,
			},
		},
		"from": map[string]interface{}{
			"email": "emails@scribly.ink",
		},
		"content": []map[string]interface{}{
			{
				"type":  "text/html",
				"value": email.Body,
			},
		},
	}
	body := bytes.Buffer{}
	if err := json.NewEncoder(&body).Encode(sendgridEmail); err != nil {
		return err
	}

	req, err := http.NewRequest("POST", s.baseURL+"/v3/mail/send", &body)
	if err != nil {
		return err
	}

	req.Header.Add("Authorization", "Bearer "+s.apiKey)
	req.Header.Add("content-type", "application/json")
	resp, err := s.client.Do(req)
	if err != nil {
		return err
	}

	if resp.StatusCode < 200 || resp.StatusCode > 299 {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("Received non-200 response from sendgrid %d - '%s'", resp.StatusCode, string(body))
	}

	return err
}

func NewSendgridClient(baseURL string, apiKey string) *SendgridClient {
	return &SendgridClient{
		client:  &http.Client{},
		baseURL: baseURL,
		apiKey:  apiKey,
	}
}
